zebra-network/src/peer_set/initialize/recent_by_ip.rs

```diff
@@ -0,0 +1,144 @@
+//! A set of IPs from recent connection attempts.
+
+use std::{
+    collections::{HashMap, VecDeque},
+    net::IpAddr,
+    time::{Duration, Instant},
+};
+
+use crate::constants;
+
+#[derive(Debug)]
+/// Stores IPs of recently attempted inbound connections.
+pub struct RecentByIp {
+    /// The list of IPs in increasing connection time order.
+    pub by_time: VecDeque<(IpAddr, Instant)>,
+
+    /// Stores IPs for recently attempted inbound connections.
+    pub by_ip: HashMap<IpAddr, usize>,
+
+    /// The maximum number of peer connections Zebra will keep for a given IP address
+    /// before it drops any additional peer connections with that IP.
+    pub max_connections_per_ip: usize,
+
+    /// The duration to wait after an entry is added before removing it.
+    pub time_limit: Duration,
+}
+
+impl Default for RecentByIp {
+    fn default() -> Self {
+        Self::new(None, None)
+    }
+}
+
+impl RecentByIp {
+    /// Creates a new [`RecentByIp`]
+    pub fn new(time_limit: Option<Duration>, max_connections_per_ip: Option<usize>) -> Self {
+        let (by_time, by_ip) = Default::default();
+        Self {
+            by_time,
+            by_ip,
+            time_limit: time_limit.unwrap_or(constants::MIN_PEER_RECONNECTION_DELAY),
+            max_connections_per_ip: max_connections_per_ip
+                .unwrap_or(constants::DEFAULT_MAX_CONNS_PER_IP),
+        }
+    }
+
+    /// Prunes outdated entries, checks if there's a recently attempted inbound connection with
+    /// this IP, and adds the entry to `by_time` if it wasn't already in `by_ip`.
+    ///
+    /// Returns true if there was a recently attempted inbound connection.
+    pub fn is_past_limit_or_add(&mut self, ip: IpAddr) -> bool {
+        let now = Instant::now();
+        self.prune_by_time(now);
+
+        let count = self.by_ip.entry(ip).or_default();
+        if *count >= self.max_connections_per_ip {
+            true
+        } else {
+            *count += 1;
+            self.by_time.push_back((ip, now));
+            false
+        }
+    }
+
+    /// Prunes entries older than `time_limit`, decrementing or removing their counts in `by_ip`.
+    fn prune_by_time(&mut self, now: Instant) {
+        while let Some((ip, time)) = self.by_time.pop_front() {
+            if now.saturating_duration_since(time) <= self.time_limit {
+                return self.by_time.push_front((ip, time));
```

Comment:
Optional Nitpick: Usually when a method returns nothing, we don't return its return value.

```suggestion
                self.by_time.push_front((ip, time));
                return;
```

Reply:
Fixed in https://github.com/ZcashFoundation/zebra/pull/7041/commits/397e70391a5b80a0097b58f740fc1a9eb8c0e7a8 and noted.